### PR TITLE
Simplify callables

### PR DESCRIPTION
**Pull request by K-ballo**
Rework `bind`, `mem_fn`, `deferred_call` implementations for readability.

Copiability requirements are enforced after these changes. In particular, movable-only `bind` objects are now movable-only instead of fake-copyable.

### CODE REVIEW

**Comment by hkaiser** [comment]
LGTM! Thanks a lot!

File: hpx/async.hpp [hunk 1]
```diff
@@ -101,12 +101,12 @@ namespace hpx { namespace detail
...
-            traits::detail::is_deferred_callable<F(Ts...)>::value,
-            hpx::future<typename util::deferred_call_result_of<F(Ts...)>::type>
+            traits::detail::is_deferred_callable<F(Ts&&...)>::value,
+            hpx::future<typename util::detail::deferred_result_of<F(Ts&&...)>::type>
         >::type
         call(BOOST_SCOPED_ENUM(launch) launch_policy, F&& f, Ts&&... ts)
         {
-            typedef typename util::deferred_call_result_of<
+            typedef typename util::detail::deferred_result_of<
                 F(Ts...)
```
Comment (sithhell) [inline hpx/async.hpp]: I am a little confused. Could you please sched some light on where we need to add ``&&`` and where not?
In particular, you have, for example, ```typename util::detail::deferred_result_of<F(Ts&&...)>::type``` in line 105 and ```typename util::detail::deferred_result_of<F(Ts...)>::type``` right here (line 109). Could you pleas elaborate on that?

File: hpx/runtime/actions/basic_action.hpp [hunk 2]
```diff
@@ -62,14 +62,12 @@ namespace hpx { namespace actions
             HPX_MOVABLE_BUT_NOT_COPYABLE(continuation_thread_function);
 
         public:
-            template <typename F_, typename ...Ts_>
             explicit continuation_thread_function(
-                    std::unique_ptr<continuation> cont,
-                    naming::address::address_type lva, F_&& f, Ts_&&... vs)
+                std::unique_ptr<continuation> cont,
+                naming::address::address_type lva, F&& f, Ts&&... vs)
```
Comment (sithhell) [inline hpx/runtime/actions/basic_action.hpp]: I am not sure if this change is correct. If I am not mistaken, ```F&&``` and ```Ts&&...``` are rvalue reference and not forwarding references in that context, correct?

File: hpx/async.hpp (hunk 1 above)
Comment (K-ballo) [inline hpx/async.hpp]: Parameter types in function types are decayed. When doing perfect forwarding, forgetting the `&&` means that for rvalues qualifiers will be lost, and things might spontaneously turn into pointers. They should all be `Ts&&...`, I missed this one.

File: hpx/runtime/actions/basic_action.hpp (hunk 2 above)
Comment (K-ballo) [inline hpx/runtime/actions/basic_action.hpp]: That's correct, this constructor does not do deduction anymore (but it is still forwarding). `F` and `Ts` are the exact types deduced earlier, there is no need to re-deduce them as the only cause for a mismatch would be bad code (which we'd now see at compile time).

Comment (sithhell) [inline hpx/runtime/actions/basic_action.hpp]: Ok, makes sense.

File: hpx/async.hpp (hunk 1 above)
Comment (sithhell) [inline hpx/async.hpp]: On 11/23/2015 01:00 PM, Agustín Bergé wrote:
> In hpx/async.hpp
> <https://github.com/STEllAR-GROUP/hpx/pull/1875#discussion_r45595802>:
> 
>>          >::type
>>          call(BOOST_SCOPED_ENUM(launch) launch_policy, F&& f, Ts&&... ts)
>>          {
>> -            typedef typename util::deferred_call_result_of<
>> +            typedef typename util::detail::deferred_result_of<
>>                  F(Ts...)
> 
> Parameter types in function types are decayed. When doing perfect
> forwarding, forgetting the |&&| means that for rvalues qualifiers will
> be lost, and things might spontaneously turn into pointers. They should
> all be |Ts&&...|, I missed this one.

Good to know, thanks.

**Pull request merged**